### PR TITLE
Add GaussianProcess layer.

### PR DESCRIPTION
**Pull request by dustinvtran**
@markvdw @dusenberrymw: Can you two review the pull request?

cc'ing for interest: @danijar @sguada @dmrd @agane @murphyk

Nothing is set in stone, so I'm open to thoughts on both API and implementation. Some limitations for simplicity:

+ Inputs must have 1-D batch dimension, not unlike most Keras layers.
+ There is only one numerical stability trick, which is perturbing the diagonals by epsilon before any calls to tf.cholesky. Any performance / numerical tips are welcome.

To implement ExponentiatedQuadratic, I copied Edward1's implementation (http://edwardlib.org/api/ed/rbf; https://github.com/blei-lab/edward/pull/596), and which is based on GPflow's.

### CODE REVIEW

**Comment by afrozenator** [comment]
@dustinvtran - let me know when this is ready. Thanks!

**Comment by dustinvtran** [comment]
> @murphyk 
> For the GP kernels, can you reuse existing TFP code?
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/positive_semidefinite_kernels/exponentiated_quadratic.py
( I find your implementation much easier to read, but I assume there are good reasons for all the TFP boilerplate code?)

Yes, I think you can pass in a TFP kernel as, e.g., `covariance_fn = lambda x1, x2: TFPKernel(parameters_here).matrix(x1, x2)`.

It looks like they're solving very general cases like any number of batch dimensions vs event dimensions, and all ways to pass inputs to a kernel function. For simplicity, I like the current signature which is that the GP accepts any covariance function taking two inputs and spitting out a matrix. That can use TFP kernels or another implementation. If the complexity necessitates it, we can explicitly restrict the input to be a TFP kernel in the future.

**Comment by murphyk** [comment]
I see. So all their complexity is due to batch and event dimensions. I
always found that very confusing. I like your interface better: data input
is just N1*D and N2*D, output is D*D matrix, just like in the textbooks.

File: tensor2tensor/layers/bayes.py [hunk 1]
```diff
@@ -41,6 +42,50 @@ def softplus():  # alias, following tf.keras.constraints
...
+    Returns:
+      Tensor of shape [batch_x1, batch_x2].
+    """
+    x1 = x1 / self.lengthscale
+    x2 = x2 / self.lengthscale
+    x1_squared = tf.reduce_sum(tf.square(x1), list(range(1, len(x1.shape))))
+    x2_squared = tf.reduce_sum(tf.square(x2), list(range(1, len(x2.shape))))
+    square = (tf.reshape(x1_squared, [-1, 1]) +
+              tf.reshape(x2_squared, [1, -1]) -
+              2 * tf.matmul(x1, x2, transpose_b=True))
```
Comment (dusenberrymw) [inline tensor2tensor/layers/bayes.py]: It seems like the docs and part of this function aim to have a kernel that supports input tensors with dims > 2, but this matmul won't work for that case, right?  Imagine x1 is shape [batch_size_x1, dim1, dim2] and x2 is shape [batch_size_x2, dim1, dim2].  This would transpose the latter two axes of x2, and attempt an element-wise multiplication across the batch axis, but that would fail unless batch_size_x1 == batch_size_x2.  We could use a tensor dot product here, such as `tf.tensordot(x1, tf.transpose(x2), axes=len(x1.shape)-1)`, which would yield an output with shape `[batch_size_x1, batch_size_x2]` regardless of the number of tensor dimensions per example.

Comment (dustinvtran) [reply tensor2tensor/layers/bayes.py]: Good pt. For now, raised NotImplementedError for dims > 2.

File: tensor2tensor/layers/bayes.py [hunk 2]
```diff
@@ -272,6 +317,146 @@ def build(self, input_shape):
...
+      covariance_matrix = self.covariance_fn(inputs, inputs)
+      covariance_matrix = tf.matrix_set_diag(
+          covariance_matrix,
+          tf.matrix_diag_part(covariance_matrix) + tf.keras.backend.epsilon())
+      # Tile locations so output has shape [units, batch_size]. Covariance will
+      # broadcast to [units, batch_size, batch_size], and we perform
+      # shape manipulations to get a random variable over [batch_size, units].
+      loc = tf.tile(loc[tf.newaxis], [self.units] + [1] * len(loc.shape))
+      return make_multivariate_normal_units(
+          loc=loc, covariance_matrix=covariance_matrix)
```
[the remark this answers is not in the document]
Comment (dustinvtran) [reply tensor2tensor/layers/bayes.py]: Done.

File: tensor2tensor/layers/bayes.py [hunk 3]
```diff
@@ -41,6 +42,50 @@ def softplus():  # alias, following tf.keras.constraints
...
+      x2: Tensor of shape [batch_x2, ...]. Slices along the batch axis denote an
+        individual input passed to the kernel function. It is computed pairwise
+        with each input sliced from x1.
+
+    Returns:
+      Tensor of shape [batch_x1, batch_x2].
+    """
+    x1 = x1 / self.lengthscale
+    x2 = x2 / self.lengthscale
+    x1_squared = tf.reduce_sum(tf.square(x1), list(range(1, len(x1.shape))))
```
[the remark this answers is not in the document]
Comment (dustinvtran) [reply tensor2tensor/layers/bayes.py]: Acknowledged.

File: tensor2tensor/layers/bayes.py [hunk 4]
```diff
@@ -41,6 +42,50 @@ def softplus():  # alias, following tf.keras.constraints
...
+    x2_squared = tf.reduce_sum(tf.square(x2), list(range(1, len(x2.shape))))
+    square = (tf.reshape(x1_squared, [-1, 1]) +
+              tf.reshape(x2_squared, [1, -1]) -
```
[the remark this answers is not in the document]
Comment (dustinvtran) [reply tensor2tensor/layers/bayes.py]: Done.

**Comment by afrozenator** [comment]
thanks @dustinvtran !
